duckduckgo/iOS
# [DuckPlayer] - Base Implementation

Task/Issue URL: https://app.asana.com/0/1204099484721401/1207593922684284/f
Tech Design URL: https://app.asana.com/0/1204099484721401/1207530480716552/f

**Description**:
- Implements base DuckPlayer functionality based on macOS code
- Abstract navigation logic and Adds a YoutubeNavigationHandler to TabViewController that is used when DuckPlayer is enabled
- Adds App Settings to Manage DuckPlayer State
- Adds Tests for the Navigation Handler to ensure redirect works as expected and completion calls are always called

**Steps to test this PR**:
1.  Make yourself internal User
2.  Go to Settings > DuckPlayer and Select 'Always'
3. Navigate to youtube.com and tap on a video
4. Confirm video loads in DuckPlayer
5. Go Back
6. Confirm you're taken back to the Main YouTube page
7. Enter a Duck URL in the Omnibar, like. duck://player/0HZPyhnErFo
8. Confirm video is loaded in DuckPlayer
9. Go back to settings and disable DuckPlayer
10. Confirm YouTube video links no longer load in Duck player

⚠️  Known issues / Work in progress

This is the first of several PRs and is still missing features

- AlwaysASK mode does not work
- Youtube Video overlays do not work
- DuckPlayer UI uses macOS HTML views (mobile views are WIP)
- DuckPlayer UI buttons do not work
- Settings Design is not fully implemented



---
###### Internal references:
[Software Engineering Expectations](https://app.asana.com/0/59792373528535/199064865822552)
[Technical Design Template](https://app.asana.com/0/59792373528535/184709971311943)
